mttl/config.py

```diff
@@ -1,89 +1,140 @@
+        if default != last_default:
+            self.defaults[cls.__name__] = default
+
+    @property
+    def default(self):
+        return next(iter(self.defaults.values())) if len(self.defaults) == 1 else self
+
+    def __repr__(self):
+        if len(self.defaults) == 1:
+            return repr(self.default)
+        return f"MultiDefaultValue({self.defaults})"
+
+
+def dataclasses_union(*args: Type[dataclass]) -> List:
+    """
+    Create a new dataclass that is the union of the fields of the input dataclasses.
+    """
+    new_fields = {}
+    for c in args:
+        for f in fields(c):
+            if f.name not in new_fields:
+                value = (
+                    f.type,
+                    field(default=MultiDefaultValue(c, f.name, f.type, f.default)),
                 )
+                new_fields[f.name] = value
+            else:
+                value = new_fields[f.name][1]
+                value.default.update(c, f.default, f.type)
 
-        if kwargs:
-            overwrite_logs += self.update_kwargs(
-                kwargs, raise_error=raise_error, silent=silent
-            )
+    # now we need to set the default values for all the fields in which there is only one default value
+    for k in new_fields:
+        value = new_fields[k][1]
+        value.default = value.default.default
 
-        # setup logging to the output dir
-        try:
-            setup_logging(self.output_dir)
-        except Exception as e:
-            if raise_error:
-                raise ValueError("Error setting up logging") from e
-            elif not silent:
-                logger.warning(f"Error setting up logging to {self.output_dir}")
+    return [(k,) + v for k, v in new_fields.items()]
 
-        # log the overwrites
-        for log in overwrite_logs:
-            logger.warning(log)
 
-        self.post_init(silent=silent)
+def create_config_class_from_args(config_class: dataclass, args: "Args"):
+    """
+    Load a dataclass from the arguments.
+    """
+    kwargs = {}
 
-    def post_init(self, silent=False):
-        if self.attn_implementation == "eager" and self.pack_sequences:
-            logger.warning(
-                "Eager attention is not compatible with packed sequences"
-                + ", tokens across examples will not be masked"
-            )
-        elif self.attn_implementation == "flash_attention_2" and self.pack_sequences:
-            logger.warning(
-                "The wrapper we provide for flash attention 2 may not behave as expected for"
-                + " some models. Please make sure you test the model with packed sequences"
-            )
+    for f in fields(config_class):
+        if hasattr(args, f.name):
+            value = getattr(args, f.name)
+
+            # if the field is not a multi-default, we override it (it was either default or set by the user)
+            if not isinstance(value, MultiDefaultValue):
+                kwargs[f.name] = value
```

Comment:
So non overwritten values are always of type `MultiDefaultValue` ?

Reply:
yeah exactly, given that it's a MultiDefaultValue, we don't pass it to the dataclass i.e. the user has not set it, in this case the dataclass will use its own default